kaiax/auction/impl/api.go

```diff
@@ -43,13 +55,87 @@ func newAuctionAPI(a *AuctionModule) *AuctionAPI {
 	return &AuctionAPI{a: a}
 }
 
-// TODO-kaiax: replace this with a correct implementation
-func (a *AuctionAPI) SubmitBid(input hexutil.Bytes) (common.Hash, error) {
-	bid := new(auction.Bid)
-	err := rlp.DecodeBytes(input, bid)
-	if err != nil {
-		return common.Hash{}, err
+type RPCOutput map[string]any
+
+// BidInput is the same format with `BidData`, replacing `[]byte` type to `hexutil.Bytes`
+type BidInput struct {
+	TargetTxRaw   hexutil.Bytes  `json:"targetTxRaw"`
+	TargetTxHash  common.Hash    `json:"targetTxHash"`
+	BlockNumber   uint64         `json:"blockNumber"`
+	Sender        common.Address `json:"sender"`
+	To            common.Address `json:"to"`
+	Nonce         uint64         `json:"nonce"`
+	Bid           *big.Int       `json:"bid"`
+	CallGasLimit  uint64         `json:"callGasLimit"`
+	Data          hexutil.Bytes  `json:"data"`
+	SearcherSig   hexutil.Bytes  `json:"searcherSig"`
+	AuctioneerSig hexutil.Bytes  `json:"auctioneerSig"`
+}
+
+func toBid(bidInput BidInput) *auction.Bid {
+	bidData := auction.BidData{
+		TargetTxHash:  bidInput.TargetTxHash,
+		BlockNumber:   bidInput.BlockNumber,
+		Sender:        bidInput.Sender,
+		To:            bidInput.To,
+		Nonce:         bidInput.Nonce,
+		Bid:           bidInput.Bid,
+		CallGasLimit:  bidInput.CallGasLimit,
+		Data:          bidInput.Data,
+		SearcherSig:   bidInput.SearcherSig,
+		AuctioneerSig: bidInput.AuctioneerSig,
+	}
+	return &auction.Bid{BidData: bidData}
+}
+
+func toTx(targetTxRaw []byte) (*types.Transaction, error) {
+	if 0 < targetTxRaw[0] && targetTxRaw[0] < 0x7f {
+		targetTxRaw = append([]byte{byte(types.EthereumTxTypeEnvelope)}, targetTxRaw...)
 	}
+	tx := new(types.Transaction)
+	if err := rlp.DecodeBytes(targetTxRaw, tx); err != nil {
+		return nil, err
+	}
+	return tx, nil
+}
 
-	return a.a.bidPool.AddBid(bid)
+func (api *AuctionAPI) SubmitBid(ctx context.Context, bidInput BidInput) RPCOutput {
+	//  1. directly send traget transaction (target tx can be empty)
+	if len(bidInput.TargetTxRaw) > 0 {
+		targetTx, errTxDecode := toTx(bidInput.TargetTxRaw)
+		if errTxDecode != nil {
+			return makeRPCOutput(common.Hash{}, errTxDecode, nil, nil)
+		}
+		errTargetTxSend := api.a.Backend.SendTx(ctx, targetTx)
+		// ignore `nonce too low` error against target tx validation
```

Comment:
why can we ignore the nonce too low?

Reply:
Because the bid might already be sent. In this case I thought this error could be ignored. Any other point of view?

Reply:
In that case, the returned error will be `known transaction: 0x...`.

Reply:
Thanks for pointing that out. The error is triggered only if the transaction is in the pool. Similarly, the current error ("nonce too low") occurs only if the transaction has already been processed.

Oh, if the tx has already been processed, the corresponding bid must be invalid. In that case, we can remove the latter scenario. :)

Reply:
Yes, nonce too low shouldn't be ignored since we need to discard a bid, but know transaction can be ignored.

Reply:
Updated.